src/Compilers/CSharp/Portable/Parser/LanguageParser_Patterns.cs

```diff
@@ -102,6 +102,132 @@ private PatternSyntax ParsePattern()
+        // or `out`, or arises while parsing the first element of a tuple literal (in which case the tokens are
+        // preceded by `(` or `:` and the identifier is followed by a `,`) or a subsequent element of a tuple literal.
+        // If the following token is among this list, or an identifier in such a context,
+        // then the *type-argument-list* is retained as part of the *simple-name*, *member-access* or
+        // *pointer-member-access* and any other possible parse of the sequence of tokens is discarded.
+        // Otherwise, the *type-argument-list* is not considered to be part of the *simple-name*,
+        // *member-access* or *pointer-member-access*, even if there is no other possible parse of the
+        // sequence of tokens. Note that these rules are not applied when parsing a *type-argument-list*
+        // in a *namespace-or-type-name* (§3.8).
+        //
+        private ExpressionSyntax ParseExpressionOrDeclaration(ParseTypeMode mode, MessageID feature, bool permitTupleDesignation)
+        {
+            if (IsPossibleDeclarationExpression(mode, permitTupleDesignation))
+            {
+                TypeSyntax type = this.ParseType(mode);
+                var designation = ParseDesignation();
+                if (feature != 0)
+                {
+                    designation = CheckFeatureAvailability(designation, feature);
+                }
+
+                return _syntaxFactory.DeclarationExpression(type, designation);
+            }
+            else
+            {
+                return this.ParseSubExpression(Precedence.Expression);
+            }
+        }
+
+        private bool IsPossibleDeclarationExpression(ParseTypeMode mode, bool permitTupleDesignation)
+        {
+            if (this.IsInAsync && this.CurrentToken.ContextualKind == SyntaxKind.AwaitKeyword)
+            {
+                // can't be a declaration expression.
+                return false;
+            }
+
+            var resetPoint = this.GetResetPoint();
+            try
+            {
+                SyntaxToken lastTokenOfType;
+                switch (ScanType(out lastTokenOfType))
+                {
+                    case ScanTypeFlags.PointerOrMultiplication:
+                        if (mode == ParseTypeMode.FirstElementOfPossibleTupleLiteral || mode == ParseTypeMode.AfterTupleComma)
+                        {
+                            // Tuples cannot contain pointer types because pointers may not be generic type arguments.
+                            return false;
+                        }
+                        break;
+
+                    case ScanTypeFlags.NotType:
+                        return false;
+                }
+
+                // check for a designation
+                if (!ScanDesignation(permitTupleDesignation && (lastTokenOfType.IsVar() || IsPredefinedType(lastTokenOfType.Kind))))
```

Comment:
Probably not, as we'll want `A.var(x, y)` to be an invocation expression. I'll revise it.